[알린] Vending step8 PR

jinios opened the pull request:
거의 일주일만에 보내는 것 같은 PR입니다...ㅠㅠ

jinios committed:
Feat: Add PieGraphView

jinios committed:
Feat: Add extension for getting radians from degrees

jinios committed:
Feat: Add PieGraphView and add one sector shape path in draw()

jinios committed:
Refactor: Declare computed property

jinios committed:
Feat: Add features in draw()
- Draw multiple sectors
- Declare new computed property - var endAngles: [Beverage: CGFloat]?

jinios committed:
Feat: Draw pie graph
Add feature - drawing pie graph

jinios committed:
Update README - step8

jinios committed:
refactor- Rename variables

jinios committed:
Feat: Add draw Text- beverage label
- Added methods: makeText(), coordinatesfromCentralAngle(), textRect()
- makeText(): Make and set custom String with specific style and attributes
- coordinatesfromCentralAngle(): Get CGpoint between central angle and arc of a pie by using trigonometric functions
- textRect(): Make CGRect for drawing beverage lable

jinios committed:
Fix: Draw Text on the pie graph
- Modify and Add functions for calculating the origin of CGRect of labels on the graph
- Modify and Add properties - arrangedHistory, endAngles, endRadians
- Refactor - Replace constants into declared variables

jinios committed:
Update README - step8

jinios committed:
Practice and examples for drawing graph and texts

godrm commented on VendingMachineApp/VendingMachineApp/PieGraphView.swift:
```diff
@@ -0,0 +1,122 @@
+    }
+    var radius: CGFloat {
+        return min(self.bounds.width, self.bounds.height) / 2
+    }
```
뷰 객체더라도 내부 전용 속성이라면 private을 꼭 붙이세요.

godrm commented on VendingMachineApp/VendingMachineApp/PieGraphView.swift:
```diff
@@ -0,0 +1,122 @@
+        return self.bounds.height / 2
+    }
+    var centerPoint: CGPoint {
+        return CGPoint(x: centerX, y: centerY)
```
centerPoint만 있으면 centerX나 centerY가 중복되는게 아닐까요?

jinios replied:
처음쓸때는 CGFloat이 익숙하지 않아서 바로 보면서 쓰기 편하라고 일부러 놔뒀는데, 지금은 편해져서 지우겠습니다!

godrm commented on VendingMachineApp/VendingMachineApp/PieGraphView.swift:
```diff
@@ -0,0 +1,122 @@
+    }
+
+    var endAngles: [CGFloat]? {
+        let angles: [String: CGFloat]? = arrangedHistory?.mapValues({CGFloat(CGFloat($0) / CGFloat((historyData?.count)!) * 360)
```
여기 CGFloat으로 바꾸는게 여러번 나오는데 모두 필요한걸까요? 
그리고 강제 형변환 코드가 있네요. 제거해보세요.

jinios replied:
알려주신대로 나누는 수 중에 `historyData?.count`를 `Double`로 변경하고, 
360은 `let degree`변수, 타입은 `Double`로 선언했지만 `$0`이 위의 arrangedHistory변수에서 가져오는 `Int`값이라 `$0 / historyData.count` 에서 연산자 에러가 나서.. 어차피 다 바꿔줘야 할 것 같습니다. 이렇게되면 아래와 같이 변경되는데, CGFloat으로 형변환하는 부분이 일부 Double로 바뀐거 이외에는 달라진게 없는 것 같습니다. 
변경한 코드 : `CGFloat(Double($0) / Double(history.count) * degree).degreesToRadians` 
mapValues 내부에서 각각의 항목을 형변환해서 쓰지 않는 방법은 위의 arrangedHistory변수타입을 `String:Double`로 변경하는 방법이 있지만, 이 방법이 크게 효과적이라고 생각하지는 않는데, 어떻게 접근하는게 좋을까요 ?_?

godrm commented on VendingMachineApp/VendingMachineApp/PieGraphView.swift:
```diff
@@ -0,0 +1,122 @@
+    private func drawText(text: String, startArc: CGFloat, arc: CGFloat) -> CGFloat {
+        let endArc = startArc + arc
+        let angle = (startArc + endArc) / 2
+        return angle
```
text 매개변수는 사용하지 않는것 같은데요? 
함수 이름도 코드 내용이나 의도를 표현하지 못하는 것 같습니다.

godrm commented on VendingMachineApp/VendingMachineApp/PieGraphView.swift:
```diff
@@ -0,0 +1,122 @@
+        return path
+    }
+
+    private func makeText(_ myText: String) -> NSAttributedString {
```
NSAttributedString 를 만들어주는 함수 의도를 좀 더 반영하는 이름으로 표현해보세요.

godrm commented on VendingMachineApp/VendingMachineApp/PieGraphView.swift:
```diff
@@ -0,0 +1,122 @@
+    }
+
+    private func coordinatesfromCentralAngle(_ angle: CGFloat) -> CGPoint {
+        let halfR = self.radius / 2
```
줄여쓰기 보다는 Radius를 그대로 표현하는게 좋겠습니다.

godrm commented on VendingMachineApp/VendingMachineApp/PieGraphView.swift:
```diff
@@ -0,0 +1,122 @@
+        var startArc = CGFloat(0)
+        var index = 0
+        for list in radiansData.keys {
+            let endArc = radiansData[list]!
```
여기도 옵셔널을 그대로 언래핑하네요.

godrm reviewed (commented):
사소한 것들을 챙겨서 디테일을 놓지지 않도록 해보세요.
피드백을 확인해보세요.

jinios committed:
Refactor: rename and modify variables, functions and parameters

jinios committed:
Refactor: Delete endAngles variable - unused

jinios committed:
Refactor: Add history variable
- Add history variable to delete forced unwrapping of historyData, arrangedHistory and endRadians

jinios commented:
- 옵셔널 변수를 바인딩해서 코드에서 강제언래핑 하는 부분을 제거했습니다.
- 어색한 변수명과 함수이름, 파라미터를 수정했습니다!

godrm commented:
마무리하러 갑시다!
